### PR TITLE
dotnet recipe draft

### PR DESCRIPTION
**Pull request by acesnik**
Checklist

- [x] License file is packaged
- [x] Source is from official source
- [ ] Package does not vendor other packages
- [x] Build number is 0
- [ ] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

### CODE REVIEW

**Comment by mariusvniekerk** [comment]
Can you put the recipe in its own directory please

File: recipes/example/meta.yaml [hunk 1]
```diff
@@ -1,83 +1,23 @@
...
-  #  `conda install openssl -c conda-forge``
+  url: https://download.visualstudio.microsoft.com/download/pr/647f8505-3bf0-48c5-ac0f-3839be6816d7/d0c2762ded5a1ded3c79b1e495e43b7c/dotnet-sdk-{{ version }}-linux-x64.tar.gz
+  sha256: '{{ sha256 }}'
 
 build:
-  # Uncomment the following line if the package is pure python and the recipe is exactly the same for all platforms.
-  # It is okay if the dependencies are not built for all platforms/versions, although selectors are still not allowed.
-  # See https://conda-forge.org/docs/meta.html#building-noarch-packages for more details.
-  # noarch: python
+  skip: True # [win]
```
Comment (dhirschfeld) [inline recipes/example/meta.yaml]:
```suggestion
```

File: recipes/example/meta.yaml [hunk 2]
```diff
@@ -1,83 +1,23 @@
...
-  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz
-  # If getting the source from GitHub remove the line above
-  # uncomment the line below and modify as needed
-  # url: https://github.com/simplejson/{{ name }}/archive/{{ version }}.tar.gz
-  sha256: d58439c548433adcda98e695be53e526ba940a4b9c44fb9a05d92cd495cdd47f
-  # sha256 is the prefered checksum -- you can get it for a file with:
-  #  `openssl sha256 <file name>`.
-  # You may need the openssl package, available on conda-forge
-  #  `conda install openssl -c conda-forge``
+  url: https://download.visualstudio.microsoft.com/download/pr/647f8505-3bf0-48c5-ac0f-3839be6816d7/d0c2762ded5a1ded3c79b1e495e43b7c/dotnet-sdk-{{ version }}-linux-x64.tar.gz
```
Comment (dhirschfeld) [inline recipes/example/meta.yaml]: May as well support all platforms:
```
{% set platform = "linux" %}  # [linux]
{% set platform = "osx" %}    # [osx]
{% set platform = "win" %}    # [win]
{% set ext= "tar.gz" %}       # [not win]
{% set ext= "zip" %}          # [win]
```
```yaml
    url: https://dotnetcli.azureedge.net/dotnet/Sdk/{{ version }}/dotnet-sdk-{{ version }}-{{ platform }}-x64.{{ ext }}
```
I think the above is a better url to use (pulled from [`dotnet-install.ps1 `](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-install-script))

File: recipes/example/meta.yaml [hunk 3]
```diff
@@ -1,83 +1,23 @@
...
-  # Remember to specify the license variants for BSD, Apache, GPL, and LGLP.
-  # Prefer the short version, e.g: GPL-2.0 instead of GNU General Public License version 2.0
-  # See https://opensource.org/licenses/alphabetical
-  license: MIT
-  # The license_family, i.e. "BSD" if license is "BSD-3-Clause". (optional)
-  license_family: MIT
-  # It is strongly encouraged to include a license file in the package,
-  # (even if the license doesn't require it) using the license_file entry.
-  # See http://conda.pydata.org/docs/building/meta-yaml.html#license-file
-  license_file: LICENSE.txt
```
Comment (dhirschfeld) [inline recipes/example/meta.yaml]: You'll need to specify the `LICENSE.txt` file which is included in the source

File: recipes/example/meta.yaml [hunk 4]
```diff
@@ -1,83 +1,23 @@
...
-    # if your project compiles code (such as a C extension) then add the required compilers as separate entries here.
-    # compilers are named 'c', 'cxx' and 'fortran'.
-    - {{ compiler('c') }}
-  host:
-    - python
-    - pip
-  run:
-    - python
+  script: 
+    - tar xvf dotnet-sdk-{{ version }}-linux-x64.tar.gz
```
Comment (dhirschfeld) [inline recipes/example/meta.yaml]: I think this needs to be unpacked somewhere meanginful to conda?
On Windows the contents of the archive are:
```
root/
│   dotnet.exe
│   LICENSE.txt
│   ThirdPartyNotices.txt
│
├───host
│   └───fxr
│       └───2.2.4
│               hostfxr.dll
├───sdk
│   └───2.2.203
│       │ 
│     <SNIP>
│
└───shared
    ├───Microsoft.AspNetCore.All
    │   └───2.2.4
    ├───Microsoft.AspNetCore.App
    │   └───2.2.4
    └───Microsoft.NETCore.App
        └───2.2.4
```

Comment (dhirschfeld) [inline recipes/example/meta.yaml]: e.g. on linux:
```
mkdir -p $PREFIX/dotnet && tar xvf dotnet-sdk-{{ version }}-{{ platform }}-x64.tar -C $PREFIX/dotnet
```
...though you'd want to have `$PREFIX/dotnet` on your path after activating the environment

Comment (acesnik) [reply recipes/example/meta.yaml]: Hmm, what location would be meaningful to conda?

**Comment by acesnik** [comment]
@dhirschfeld I'm not really sure where to go from here. It looks like it built in Appveyor, which is a plus.

**Comment by acesnik** [comment]
Oh wait, I found the error, finally:
```
RuntimeError: SHA256 mismatch: 'e5263822306be828a5923a39ec334ae6cf174a5bd4d1578a647d8cd81fbc9c72' != 'd1e2368b1335a6a5f496b887950ad5da3d85783ec76f74a663214989817bb497'
```

**Comment by acesnik** [comment]
@dhirschfeld Could you advise on what's going wrong with the Linux build? Thanks!

**Comment by dhirschfeld** [comment]
@acesnik - I'll try and take a look this week

**Comment by msarahan** [comment]
I don't see how this is getting "installed" at all.  Even on Appveyor, it's passing solely because dotnet is already installed elsewhere and on PATH: 
  The package is completely empty: https://ci.appveyor.com/project/conda-forge/staged-recipes/builds/23960835#L212

I think you might need to spend some quality time with the docs at https://python-packaging-tutorial.readthedocs.io/en/latest/conda.html# - there's a youtube video if you prefer: https://www.youtube.com/watch?v=xiI1i525ljE

This is also appears to be a binary repack.  We strongly prefer to build from source wherever feasible.  This helps us ensure that we understand our platform compatibility (e.g. which specific versions of Linux/Mac can run our software.  Are there source packages available for this?

@dhirschfeld asked on the gitter chat whether this package should be considered a top-level "language" package, like R or python.  I'm not sure.  If you anticipate building out packages that use dotnet, and these packages are sensitive to which version of dotnet was used to build them, then yes.  On the other hand, if there's no compilation or otherwise that renders a package specific to the dotnet that was used to build it, then I think it's unnecessary, and you can treat this central dotnet package as just another dependency.

**Comment by dhirschfeld** [comment]
It should be possible to build it yourself since it's open source but it would certainly be more difficult

IIUC, the dotnet binary wants to find it's "standard-library" in the sdk/shared folders below the `dotnet`  binary so it would seem to make sense to install everything to `$PREFIX/dotnet`. If you do that though the `dotnet` binary wouldn't be on the path.

Is there some way to add `$PREFIX/dotnet` to the path when the environment is activated? And remove it when it is deactivated?  Is the officially blessed way of doing this to put a script in `$PREFIX/etc/conda/activate.d/` to modify the PATH?

**Comment by acesnik** [comment]
Hmm, good catch that the package is empty. I'm new to conda/python packaging. Thanks for the pointer to the docs.

I know how to install dotnet from binary and from aptitude, but I don't know how to build it from source. 
- Homebrew is using the packaged distributions, too. https://github.com/Homebrew/homebrew-cask/blob/master/Casks/dotnet-sdk.rb 
- Here are the libraries. I don't see a quick guide for building it into dotnet. https://github.com/dotnet/corefx
- They don't offer the source code in their releases:
image
- For the record, I asked if the dotnet team wanted to contribute the package to `conda`, and they respectfully declined. https://github.com/dotnet/core/issues/2518

That's an interesting question regarding whether it should be a "top-level" language. My subfield of bioinformatics (proteomics) uses C# quite a bit. It not the most common language out there, as you know. That said, it can be used to build and run packages, so it's most akin to `javac` and `java` in one. I do intend to use it both to build and run packages in my current use case.

@dhirschfeld I don't really understand your points about putting it on the PATH. Wouldn't that vary by operating system? I tried out that `$PREFIX/dotnet` suggestion, and it didn't pass the tests, as you seem to have expected. Do you have an example script that would go into the `activate.d` folder? Maybe there'll be more info about that in the docs...

Cheers,

AC

**Comment by acesnik** [comment]
I don't have time for this one right now and found a workaround for my project. 👋

**Pull request closed without merging**